File: V1/frontend/src/services/accounts.ts
import axios from 'axios';
import { API_URL } from '../constants/api';
import { AccountAddType, AccountDeleteType, AccountType, AccountUpdateType } from '../types/Account';
import { ErrorResponseType } from '../types/Response';
export async function getAccounts(user_id: string): Promise<{ accounts: AccountType[] } | ErrorResponseType> {
  try {
    const { data, status } = await axios.get(API_URL + '/account?user_id=' + user_id);

    if (status === 200) {
      console.info('Accounts lists fetched');
      return { accounts: data };
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error.message);
      return { error: error.message, statusCode: error.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function addAccount(new_account: AccountAddType): Promise<AccountType | ErrorResponseType> {
  const body = {
    user_id: new_account.user_id,
    bank_name: new_account.bank_name,
    account_number: new_account.account_number,
    account_type: new_account.account_type,
    balance: Number(new_account.balance)
  };

  try {
    const { data, status } = await axios.post(API_URL + '/account', body);

    if (status === 201) {
      console.info('Account added successfully');
      return data;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error);
      return { error: error.response.data, statusCode: error.response.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function updateAccount(update_account: AccountUpdateType): Promise<AccountType | ErrorResponseType> {
  const body = {
    id: update_account.id,
    user_id: update_account.user_id,
    bank_name: update_account.bank_name,
    account_number: update_account.account_number,
    account_type: update_account.account_type,
    balance: Number(update_account.balance)
  };

  try {
    const { data, status } = await axios.put(API_URL + '/account', body);

    if (status === 200) {
      console.info('Account updated successfully');
      return data;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error);
      return { error: error.response.data, statusCode: error.response.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function deleteAccount(deleteAccount: AccountDeleteType): Promise<boolean | ErrorResponseType> {
  const body = {
    id: deleteAccount.id,
    user_id: deleteAccount.user_id
  };
  try {
    const { status } = await axios.delete(API_URL + '/account', { data: body });

    if (status === 204) {
      console.info('Account deleted');
      return true;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error.message);
      return { error: error.message, statusCode: error.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function getAccountType(): Promise<string[] | ErrorResponseType> {
  try {
    const { data, status } = await axios.get(API_URL + '/account-type');

    if (status === 200) {
      console.info('AccountTypes fetched');

      return data;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error.message);
      return { error: error.message, statusCode: error.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

File: backend/app/routers/accounts.py
from typing import List

from app.core.controller.Account import AccountController
from app.core.schema.Accounts.Request import (
    CreateRequest,
    DeleteRequest,
    UpdateBalanceRequest,
    UpdateRequest,
)
from app.core.schema.Accounts.Response import Response
from app.core.schema.Responses.Error import ErrorModel, Errors
from app.core.schema.Responses.Success import Success
from fastapi import APIRouter, status
from fastapi.responses import ORJSONResponse
from pymongo.errors import DuplicateKeyError

accountRouter = APIRouter(prefix="/account")
accountController = AccountController()


@accountRouter.get(
    "/{id}-{user_id}",
    response_model=Response,
    status_code=status.HTTP_200_OK,
    tags=["Account-Controller"],
    dependencies=None,
    summary="Get Account",
    description="Get a account by id",
    response_description="Successful Response",
    responses={
        Success.HTTP_200_OK.status_code: {
            "model": Response,
            "description": "Account found",
        },
        Errors.HTTP_404_NOT_FOUND.status_code: {
            "model": ErrorModel,
            "description": "Account not found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def read_todo(id: str, user_id: str) -> Response | ORJSONResponse:
    if account := await accountController.get_account(id=id, user_id=user_id):
        return account
    else:
        return Errors.HTTP_404_NOT_FOUND


@accountRouter.get(
    "/",
    response_model=List[Response],
    status_code=status.HTTP_200_OK,
    tags=["Account-Controller"],
    dependencies=None,
    summary="Get All accounts",
    description="Get all accounts fdr user",
    response_description="Successful Response",
    responses={
        Success.HTTP_200_OK.status_code: {
            "model": List[Response],
            "description": "Accounts found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def read_account_list(user_id: str) -> List[Response] | ORJSONResponse:
    if accounts := await accountController.get_all(user_id=user_id):
        return accounts
    else:
        return Errors.HTTP_404_NOT_FOUND


@accountRouter.post(
    "/",
    response_model=Response,
    status_code=status.HTTP_201_CREATED,
    tags=["Account-Controller"],
    dependencies=None,
    summary="Create Account",
    description="Create a new account in db",
    response_description="Successful Response",
    responses={
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
        Errors.HTTP_409_ALREADY_EXISTS.status_code: {
            "model": ErrorModel,
            "description": "Account already exists",
        },
        status.HTTP_201_CREATED: {
            "model": Response,
            "description": "Successfully Created",
        },
    },
)
async def create_account(account: CreateRequest) -> Response | ORJSONResponse:
    try:
        return await accountController.create_account(
            user_id=account.user_id,
            bank_name=account.bank_name,
            account_type=account.account_type,
            account_number=account.account_number,
            balance=account.balance,
        )
    except DuplicateKeyError:
        return Errors.HTTP_409_ALREADY_EXISTS


@accountRouter.put(
    "/",
    response_model=Response,
    status_code=status.HTTP_200_OK,
    tags=["Account-Controller"],
    dependencies=None,
    summary="Update Account",
    description="Update the account based on id",
    response_description="Successful Response",
    responses={
        Success.HTTP_200_OK.status_code: {
            "model": Response,
            "description": "Account updated",
        },
        Errors.HTTP_404_NOT_FOUND.status_code: {
            "model": ErrorModel,
            "description": "Account not found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def update_account(account: UpdateRequest) -> Response | ORJSONResponse:
    if updated_account := await accountController.update_account(
        id=account.id,
        user_id=account.user_id,
        bank_name=account.bank_name,
        account_type=account.account_type,
        account_number=account.account_number,
        balance=account.balance,
    ):
        return updated_account
    else:
        return Errors.HTTP_404_NOT_FOUND


@accountRouter.put(
    "/balance",
    response_model=Response,
    status_code=status.HTTP_200_OK,
    tags=["Account-Controller"],
    dependencies=None,
    summary="Update Account balance",
    description="Update the account balance based on id",
    response_description="Successful Response",
    responses={
        Success.HTTP_200_OK.status_code: {
            "model": Response,
            "description": "Account balance updated",
        },
        Errors.HTTP_404_NOT_FOUND.status_code: {
            "model": ErrorModel,
            "description": "Account not found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def update_account_balance(
    account: UpdateBalanceRequest,
) -> Response | ORJSONResponse:
    if updated_account := await accountController.update_account_balance(
        id=account.id,
        user_id=account.user_id,
        balance=account.balance,
    ):
        return updated_account
    else:
        return Errors.HTTP_404_NOT_FOUND


@accountRouter.delete(
    "/",
    response_model=None,
    status_code=status.HTTP_204_NO_CONTENT,
    tags=["Account-Controller"],
    dependencies=None,
    summary="Delete account",
    description="Delete the account based on id",
    response_description="Successful Response",
    responses={
        Success.HTTP_204_NO_CONTENT: {
            "model": None,
            "description": "Account deleted successfully",
        },
        Errors.HTTP_404_NOT_FOUND.status_code: {
            "model": ErrorModel,
            "description": "Account not found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def delete_todo(account: DeleteRequest) -> ORJSONResponse | int:
    deleted_account = await accountController.delete_account(
        id=account.id, user_id=account.user_id
    )

    if deleted_account is False:
        return Errors.HTTP_404_NOT_FOUND

    return Success.HTTP_204_NO_CONTENT
